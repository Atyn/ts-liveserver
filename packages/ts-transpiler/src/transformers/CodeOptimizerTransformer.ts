import TypeScript from 'typescript'

/*
Replaces process.env.NODE_ENV in code
*/
export default class CodeOptimizerTransformer
	implements TypeScript.CustomTransformer {
	private context: TypeScript.TransformationContext
	constructor(context: TypeScript.TransformationContext) {
		this.context = context
	}
	public transformSourceFile(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const withoutUseStrict = this.reduceUseStrict(sourceFile)
		const wihoutEmptyFunctions = this.reduceEmptyFunctions(withoutUseStrict)
		const reducedBinaryExpressions = this.reduceBinaryExpressions(
			wihoutEmptyFunctions,
		)
		return this.reduceIfStatements(reducedBinaryExpressions)
	}
	public transformBundle(node: TypeScript.Bundle): TypeScript.Bundle {
		return node
	}
	private reduceUseStrict(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] | undefined => {
			if (
				TypeScript.isExpressionStatement(node) &&
				TypeScript.isStringLiteral(node.expression) &&
				node.expression.text === 'use strict'
			) {
				return undefined
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	private reduceEmptyFunctions(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] => {
			if (
				TypeScript.isExpressionStatement(node) &&
				TypeScript.isCallExpression(node.expression) &&
				node.expression.arguments.length === 0 &&
				TypeScript.isParenthesizedExpression(node.expression.expression) &&
				TypeScript.isFunctionExpression(
					node.expression.expression.expression,
				) &&
				node.expression.expression.expression.parameters.length === 0 &&
				TypeScript.isBlock(node.expression.expression.expression.body) &&
				node.expression.expression.expression.body.statements &&
				node.expression.expression.expression.body.statements.length
			) {
				return [
					...TypeScript.visitNodes(
						node.expression.expression.expression.body.statements,
						visit,
					),
				]
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	private reduceBinaryExpressions(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] => {
			const analyzedNode = TypeScript.visitEachChild(node, visit, this.context)
			if (
				TypeScript.isBinaryExpression(analyzedNode) &&
				this.isExpressionStaticallyTruthy(analyzedNode)
			) {
				return TypeScript.factory.createTrue()
			}
			if (
				TypeScript.isBinaryExpression(analyzedNode) &&
				this.isExpressionStaticallyFalsy(analyzedNode)
			) {
				return TypeScript.factory.createFalse()
			}
			return analyzedNode
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	private reduceIfStatements(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] | undefined => {
			if (TypeScript.isIfStatement(node)) {
				// Only if
				if (
					node.expression.kind === TypeScript.SyntaxKind.TrueKeyword &&
					TypeScript.isBlock(node.thenStatement)
				) {
					return [
						...TypeScript.visitNodes(node.thenStatement.statements, visit),
					]
				}
				// Only else
				if (node.expression.kind === TypeScript.SyntaxKind.FalseKeyword) {
					if (node.elseStatement && TypeScript.isBlock(node.elseStatement)) {
						// return node.elseStatement.statements
						return [
							...TypeScript.visitNodes(node.elseStatement.statements, visit),
						]
					} else {
						return undefined
					}
				}
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	private isExpressionStaticallyTruthy(node: TypeScript.Expression): boolean {
		// The same
		if (
			TypeScript.isBinaryExpression(node) &&
			(node.operatorToken.kind === TypeScript.SyntaxKind.EqualsEqualsToken ||
				node.operatorToken.kind ===
					TypeScript.SyntaxKind.EqualsEqualsEqualsToken)
		) {
			if (
				TypeScript.isStringLiteral(node.left) &&
				TypeScript.isStringLiteral(node.right) &&
				node.left.text == node.right.text
			) {
				return true
			}
		}
		// NOT the same
		if (
			TypeScript.isBinaryExpression(node) &&
			(node.operatorToken.kind ===
				TypeScript.SyntaxKind.ExclamationEqualsEqualsToken ||
				node.operatorToken.kind ===
					TypeScript.SyntaxKind.ExclamationEqualsToken)
		) {
			if (
				TypeScript.isStringLiteral(node.left) &&
				TypeScript.isStringLiteral(node.right) &&
				node.left.text != node.right.text
			) {
				return true
			}
		}
		return false
	}
	private isExpressionStaticallyFalsy(node: TypeScript.Expression): boolean {
		if (
			TypeScript.isBinaryExpression(node) &&
			(node.operatorToken.kind === TypeScript.SyntaxKind.EqualsEqualsToken ||
				node.operatorToken.kind ===
					TypeScript.SyntaxKind.EqualsEqualsEqualsToken)
		) {
			if (
				TypeScript.isStringLiteral(node.left) &&
				TypeScript.isStringLiteral(node.right) &&
				node.left.text !== node.right.text
			) {
				return true
			}
		}
		if (
			TypeScript.isBinaryExpression(node) &&
			(node.operatorToken.kind ===
				TypeScript.SyntaxKind.ExclamationEqualsEqualsToken ||
				node.operatorToken.kind ===
					TypeScript.SyntaxKind.ExclamationEqualsToken)
		) {
			if (
				TypeScript.isStringLiteral(node.left) &&
				TypeScript.isStringLiteral(node.right) &&
				node.left.text == node.right.text
			) {
				return true
			}
		}
		return false
	}
}
