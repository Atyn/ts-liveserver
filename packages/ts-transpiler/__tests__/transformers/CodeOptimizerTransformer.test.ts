import CodeOptimizerTransformer from '../../src/transformers/CodeOptimizerTransformer'
import TypeScript from 'typescript'

describe('CodeOptimizerTransformer', () => {
	it('Create booleans of statements', async () => {
		process.env.NODE_ENV = 'development'
		const input = '"a" === "b"'
		const output = 'false'
		expect(await transformWithPlugin(input)).toBe(
			await transformWithoutPlugin(output),
		)
	})
	it('Should remove if it is the same', async () => {
		process.env.NODE_ENV = 'development'
		const input =
			'if ("astring" === "astring") { console.log("hello"); } else { console.log("never happens")}'
		const output = 'console.log("hello")'
		expect(await transformWithPlugin(input)).toBe(
			await transformWithoutPlugin(output),
		)
	})
	it('Should remove if statement if it is never the same', async () => {
		process.env.NODE_ENV = 'development'
		const input =
			'if ("something else" === "astring") { console.log("hello"); } else { console.log("always happens")}'
		const output = 'console.log("always happens")'
		expect(await transformWithPlugin(input)).toBe(
			await transformWithoutPlugin(output),
		)
	})
	it('Should empty functions in root scope', async () => {
		process.env.NODE_ENV = 'development'
		const input = '(function () { console.log("yoyo") })();'
		const output = 'console.log("yoyo")'
		expect(await transformWithPlugin(input)).toBe(
			await transformWithoutPlugin(output),
		)
	})
	it('Should combine optimizations', async () => {
		process.env.NODE_ENV = 'development'
		const input = `(function () {
			"a" === "b";
		})();`
		const output = 'false'
		expect(await transformWithPlugin(input)).toBe(
			await transformWithoutPlugin(output),
		)
	})
})

const compilerOptions: TypeScript.CompilerOptions = {
	allowJs: true,
	jsx: TypeScript.JsxEmit.React,
	checkJs: false,
	noResolve: false,
	esModuleInterop: true,
	skipLibCheck: true,
	pretty: true,
	allowUnreachableCode: true,
	target: TypeScript.ScriptTarget.ES2020,
	declaration: false,
	module: TypeScript.ModuleKind.ES2020,
	moduleResolution: TypeScript.ModuleResolutionKind.NodeJs,
	sourceMap: false,
}

const transformers: TypeScript.CustomTransformers = {
	before: [(context) => new CodeOptimizerTransformer(context)],
}

async function transformWithPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
		transformers: transformers,
	})
	return results.outputText.trim()
}

async function transformWithoutPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
	})
	return results.outputText.trim()
}
