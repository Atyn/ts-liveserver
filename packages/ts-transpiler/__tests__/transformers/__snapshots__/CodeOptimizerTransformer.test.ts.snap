// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeOptimizerTransformer Should be "true" 1`] = `"false;"`;

exports[`CodeOptimizerTransformer Should combine optimizations 1`] = `"false;"`;

exports[`CodeOptimizerTransformer Should empty functions in root scope 1`] = `"console.log(\\"yoyo\\");"`;

exports[`CodeOptimizerTransformer Should handle empty string 1`] = `
"if (\\"\\") {
    console.log(\\"hello\\");
}"
`;

exports[`CodeOptimizerTransformer Should handle false 1`] = `
"if (\\"\\") {
    console.log(\\"hello\\");
}"
`;

exports[`CodeOptimizerTransformer Should handle nested binary expressions 1`] = `"(true) === true;"`;

exports[`CodeOptimizerTransformer Should handle null 1`] = `
"if (null) {
    console.log(\\"hello\\");
}"
`;

exports[`CodeOptimizerTransformer Should handle true 1`] = `"console.log(\\"hello\\");"`;

exports[`CodeOptimizerTransformer Should remove if it is the same 1`] = `"console.log(\\"hello\\");"`;

exports[`CodeOptimizerTransformer Should remove if statement if it is never the same 1`] = `"console.log(\\"always happens\\");"`;
